cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/sourcecontrol/PullAppsOperationTest.java

```diff
@@ -0,0 +1,200 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.mockito.Mockito;
+
+public class PullAppsOperationTest {
+
+  private static final AppRequest<?> TEST_APP_REQUEST = new AppRequest<>(
+      new ArtifactSummary("name", "version"));
+
+  private static final OperationRunId runID = new OperationRunId("default", "id");
+
+  private static final Gson GSON = new Gson();
+
+  private final PullAppsRequest req = new PullAppsRequest(ImmutableSet.of("1", "2", "3", "4"));
+
+  private InMemorySourceControlOperationRunner opRunner;
+
+  @ClassRule
+  public static TemporaryFolder repositoryBase = new TemporaryFolder();
+
+  @Before
+  public void setUp() throws Exception {
+    RepositoryManagerFactory mockRepositoryManagerFactory = Mockito.mock(
+        RepositoryManagerFactory.class);
+    RepositoryManager mockRepositoryManager = Mockito.mock(RepositoryManager.class);
+    Mockito.doReturn(mockRepositoryManager).when(mockRepositoryManagerFactory)
+        .create(Mockito.any(), Mockito.any());
+    Mockito.doReturn(Paths.get("")).when(mockRepositoryManager).getRepositoryRoot();
+    Mockito.doReturn(repositoryBase.getRoot().toPath()).when(mockRepositoryManager)
+        .getRepositoryRoot();
+    Mockito.doReturn(repositoryBase.getRoot().toPath()).when(mockRepositoryManager).getBasePath();
+    Mockito.doReturn("testHash")
+        .when(mockRepositoryManager)
+        .getFileHash(Mockito.any(), Mockito.any());
+    Path tempFile = repositoryBase.newFile().toPath();
+    Files.write(tempFile, GSON.toJson(TEST_APP_REQUEST).getBytes(StandardCharsets.UTF_8));
+    Mockito.doNothing().when(mockRepositoryManager).close();
+    Mockito.doReturn(tempFile.getFileName()).when(mockRepositoryManager)
+        .getFileRelativePath(Mockito.any());
+
+    this.opRunner = new InMemorySourceControlOperationRunner(mockRepositoryManagerFactory);
+  }
+
+  @Test
+  public void testRunSuccess() throws Exception {
+    SourceControlHelper mockHelper = Mockito.mock(SourceControlHelper.class);
+    LongRunningOperationContext mockContext = Mockito.mock(LongRunningOperationContext.class);
+    PullAppsOperation operation = new PullAppsOperation(
+        this.req,
+        opRunner,
+        mockHelper
+    );
+    Set<OperationResource> gotResource = new HashSet<>();
+
+    Mockito.doReturn(runID).when(mockContext).getRunId();
+    Mockito.doAnswer(i -> {
+      gotResource.addAll(
+          (Collection<? extends OperationResource>) i.getArguments()[0]);
+      return null;
+    }).when(mockContext).updateOperationResources(Mockito.any());
+
+    OperationError err = operation.run(mockContext).get();
+
+    Assert.assertNull(err);
+    Mockito.verify(mockHelper, Mockito.times(4)).deployAppVersion(Mockito.any(), Mockito.any());
```

Comment:
I agree with the above if `run()` was a pure function, where verifying output only would work to validate the result. But we have side effects like `markVersionsLatest` that will not be covered by any output and IMHO that should be verified by checking if we are correctly using downstream units i.e `helper` interface in this case.

I can change the `run()` method to return `OperationResources` as also suggested above but I feel that has lesser value as the resources should be updated per sub-operation rather than at the operation end making the return value unused.